### PR TITLE
Refactor loading components to ScholarshipList component

### LINKED ISSUE #487: Refactor scholarship loading rendering into ScholarshipList component
Take a look at these two files:
https://github.com/beyondhb1079/s4us/blob/27a0d0e693f1e18b9298c9813260a67b7d6489e4/src/pages/UserHome.jsx#L164-L174

https://github.com/beyondhb1079/s4us/blob/27a0d0e693f1e18b9298c9813260a67b7d6489e4/src/pages/Scholarships.jsx#L146-L160

They're very similar:
- Both show a list of scholarships
- Both have a **Load More** button
- Both show [circular progress indicators](https://material-ui.com/components/progress/#circular) when loading (or loading more) scholarships

What if we update one page but forget to update the other? What if we add something similar in a different page/component? 

`ScholarshipList` only takes care of displaying the list but not the dynamics around it. Perhaps it makes sense to merge these there. Along with the `scholarships` property maybe it makes sense to add a `loadingMore` boolean property and `onLoadMore` callback property.

### PR DESCRIPTION
**Pull request by josueetcom**
- `Scholarships.list` has been updated so that it can accept `minAmount` and `maxAmount` filters instead of having to do those on its return value.
- **ScholarshipList** now supports displaying everything related to loading scholarship results, including:
  - paginated loading
  - error message displaying
  - no results (or no more results) displaying

## Motivation and Context

Fixes #487, which simplifies how we load and display scholarships, as can be evidenced by the simplification in the **Scholarships** and **UserHome** pages.

## How Has This Been Tested?

I've tested the changes on my local server and a few tests for the postProcessing filter logic

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] User visible change (users will notice UI or functional changes)

## Previewing Changes

1. Go to https://s4us-pr-489.onrender.com/scholarships.
2. Try loading more scholarships.
3. Notice slight changes in UI.

Also try logging in and making sure the UI looks correct for your dashboard page.

## Checklist:

- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: src/components/ScholarshipList.jsx [hunk 1]
```diff
@@ -35,27 +56,47 @@ function ScholarshipList({ scholarships }) {
...
+        <Box className={classes.loadMore}>
+          {onLoadMore ? (
+            <Button color="primary" onClick={onLoadMore}>
+              Load More
+            </Button>
+          ) : (
+            <Typography>End of results</Typography>
+          )}
+        </Box>
+      )}
```
Comment (josueetcom) [inline src/components/ScholarshipList.jsx:85]: Maybe this should be embedded in the Grid itself?

Comment (josueetcom) [reply src/components/ScholarshipList.jsx]: Figured it out!

File: src/pages/Scholarships.jsx [hunk 2]
```diff
@@ -33,30 +27,14 @@ const sortOptions = {
...
-  const [canLoadMore, setCanLoadMore] = useState(false);
-
   const history = useHistory();
   const location = useLocation();
 
   const params = queryString.parse(location.search, { parseNumbers: true });
 
   const setQueryParam = (index, val) => {
-    setScholarships([]);
+    // setScholarships([]); needed?
```
Comment (josueetcom) [inline src/pages/Scholarships.jsx:37]: TODO(me): Figure out if this is still needed

Comment (josueetcom) [reply src/pages/Scholarships.jsx]: It is, elsewhere though.

File: src/models/Scholarships.ts [hunk 3]
```diff
@@ -73,7 +75,10 @@ class Scholarships extends FirestoreCollection<ScholarshipData> {
       query = query.where('authorId', '==', opts.authorId);
     }
 
-    return FirestoreCollection.list(query);
+    const postProcessFilter = (s: FirestoreModel<ScholarshipData>) =>
+      s.data.amount.intersectsRange(opts.minAmount, opts.maxAmount);
```
Comment (edwingl10) [inline src/models/Scholarships.ts:79]: Nice!

File: src/components/ScholarshipList.jsx [hunk 4]
```diff
@@ -52,10 +134,12 @@ function ScholarshipList({ scholarships }) {
...
 
 ScholarshipList.propTypes = {
-  scholarships: PropTypes.arrayOf(PropTypes.object),
+  listFn: PropTypes.func,
+  noResultsNode: PropTypes.node,
 };
 ScholarshipList.defaultProps = {
-  scholarships: [],
+  listFn: undefined,
+  noResultsNode: <Typography>No scholarships found</Typography>,
```
Comment (edwingl10) [inline src/components/ScholarshipList.jsx:142]: ooo I didn't know we could set jsx as default props

**Review by edwingl10 — APPROVED** [review]

**Review by gonzalo1330 — APPROVED** [review]

File: src/components/ScholarshipList.jsx (hunk 4 above)
Comment (josueetcom) [reply src/components/ScholarshipList.jsx:142]: Yeah! The only thing tricky about it is when you need to set a `className`. In that case, I would set this `undefined` and when rendering set the value there.

**Pull request merged**